### PR TITLE
tidy: add include-what-you-use

### PR DESCRIPTION
**Pull request by fanquake**
We recently added a [`clang-tidy` job](https://github.com/bitcoin/bitcoin/blob/master/ci/test/00_setup_env_native_tidy.sh) to the CI, which generates a compilation database. We can leverage that now existing database to begin running [include-what-you-use](https://include-what-you-use.org/) over the codebase.

This PR demonstrates using a mapping_file to indicate fixups / includes that may differ from IWYU suggestions. In this case, I've added some fixups for glibc includes that I've [upstreamed changes for](https://github.com/include-what-you-use/include-what-you-use/pull/1026):
```bash
# Fixups / upstreamed changes
[
  { include: [ "<bits/termios-c_lflag.h>", private, "<termios.h>", public ] },
  { include: [ "<bits/termios-struct.h>", private, "<termios.h>", public ] },
  { include: [ "<bits/termios-tcflow.h>", private, "<termios.h>", public ] },
]
```

The include "fixing" commits of this PR:
* Adds missing includes.
* Swaps C headers for their C++ counterparts.
* Removes the pointless / unmaintainable `//for abc, xyz` comments. When using IWYU, if anyone wants to see / generate those comments, to see why something is included, it is trivial to do so (IWYU outputs them by default). i.e:
```cpp
// The full include-list for compat/stdin.cpp:
#include <compat/stdin.h>
#include <poll.h>                  // for poll, pollfd, POLLIN
#include <termios.h>               // for tcgetattr, tcsetattr
#include <unistd.h>                // for isatty, STDIN_FILENO
```

TODO:
- [ ] Qt mapping_file. There is one in the IWYU repo, but it's for Qt 5.11. Needs testing.
- [ ] Boost mapping_file. There is one in the IWYU repo, but it's for Boost 1.75. Needs testing.

I'm not suggesting we turn this on the for entire codebase, or immediately go-nuts refactoring all includes. However I think our dependency includes are now slim enough, and our CI infrastructure in place such that we can start doing this in some capacity, and just automate away include fixups / refactorings etc.

### CODE REVIEW

**Comment by jonatack** [comment]
Concept ACK

**Comment by jonatack** [comment]
It might be good to add some developer notes docs on this if it moves forward.

**Comment by ryanofsky** [comment]
Concept ACK and thanks for working on this! I've used IWYU on other projects and it was nice once you got it set up. I think ideally it would be easy to run IWYU locally, but getting it working on CI is useful, and should make it easier to run locally as well.

**Comment by laanwj** [comment]
Concept ACK

File: src/compat/stdin.cpp [hunk 1]
```diff
@@ -3,18 +3,18 @@
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
 #if defined(HAVE_CONFIG_H)
-#include <config/bitcoin-config.h>
+#include <config/bitcoin-config.h> // IWYU pragma: keep
```
Comment (MarcoFalke) [inline src/compat/stdin.cpp:6]: why is this needed?

**Review by maflcko — APPROVED** [review]

File: src/compat/stdin.cpp (hunk 1 above)
Comment (fanquake) [reply src/compat/stdin.cpp:6]: > why is this needed? (I mean the include)

In this case, the include isn't actually needed. Although in other cases we'll likely use the same pragma on this same include to suppress IWYU output, given we aren't building for all configurations (I think that's ok). I'm removing it now.

**Comment by MarcoFalke** [comment]
review ACK 9b0a13a2891641a3d12e525cee8ddddb1aa1bc73

**Review by maflcko — APPROVED** [review]

**Comment by jonatack** [comment]
Ah, ready for review -- testing.

**Review by jonatack — COMMENTED** [review]
ACK 9b0a13a2891641a3d12e525cee8ddddb1aa1bc73 reviewed changes and run CI output in https://cirrus-ci.com/task/4750910332076032

**Comment by MarcoFalke** [comment]
Going to merge this so that ppl can start experimenting with this and also use it in more places. Hopefully this will help us reduce compile time by nuking useless includes and reduce review burden by automating the review discussions on which headers to include/remove.

If there are any issues, it should be trivial to disable in CI by editing `ci/test/06_script_b.sh`.

File: ci/test/06_script_b.sh [hunk 2]
```diff
@@ -37,6 +37,8 @@ fi
 if [ "${RUN_TIDY}" = "true" ]; then
   export P_CI_DIR="${BASE_BUILD_DIR}/bitcoin-$HOST/src/"
   CI_EXEC run-clang-tidy "${MAKEJOBS}"
+  export P_CI_DIR="${BASE_BUILD_DIR}/bitcoin-$HOST/"
+  CI_EXEC "python3 ${BASE_SCRATCH_DIR}/iwyu/include-what-you-use/iwyu_tool.py src/compat src/init -p . ${MAKEJOBS} -- -Xiwyu --cxx17ns -Xiwyu --mapping_file=${BASE_BUILD_DIR}/bitcoin-$HOST/contrib/devtools/iwyu/bitcoin.core.imp"
```
[the remark this answers is not in the document]
Comment (hebasto) [reply ci/test/06_script_b.sh:41]: See #26763.
